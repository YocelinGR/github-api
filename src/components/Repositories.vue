<template>
  <div>
    <div v-for="repo in user.repositories.nodes" :key="repo.id" class="repos row">
      <div class="col-md-6">
        <a :href="repo.url" class="repoName">{{ repo.name}}</a>
        <br>
        <span v-if="repo.repositoryTopics.nodes" class="topics">
          <span v-for="topic in repo.repositoryTopics.nodes" :key="topic.id">
            {{topic.topic.name}}
          </span>
        </span>
        <span class="topics">{{ repo.createdAt }}</span>
      </div>
      <div class="col-md-3">
        <button @click="starredRepo(repo.id)" class="btn btn-light" @dblclick="removeStar(repo.id)"><i :class="repo.viewerHasStarred ? 'fas fa-star' : 'far fa-star'"></i></button>
      </div>
    </div>
  </div>
</template>

<script>
import gql from 'graphql-tag'

const overviewTag = gql`
  query user {
  user(login: "YocelinGR"){
    id
    repositories(last: 20){
      nodes{
        id
        name
        url
        url
        viewerHasStarred
        createdAt
        repositoryTopics(first: 3){
          nodes{
            id
            resourcePath
            url
            topic{
              id
              name
            }
          }
        }
      }
    }
  }
}
`;
const addStar = gql`
  mutation addStar($input: AddStarInput!){
  addStar(input: $input) {
    clientMutationId
    starrable{
      id
    }
  }
}
`;
const removingStar = gql`
  mutation removeStar($input: RemoveStarInput!){
  removeStar(input: $input) {
    clientMutationId
    starrable{
      id
    }
  }
}
`;
export default {
  data: () => ({
    user: {},
  }),
  methods: {
    starredRepo(id){
      const varStar = {
              starrableId: id,
              clientMutationId:  "MDQ6VXNlcjM5ODMzMDQ1"
            };
      this.$apollo.mutate({
        mutation: addStar,
        variables: {
            input: varStar,
        }
      })
      console.log('added');
    },
    removeStar(id){
      const varStar = {
              starrableId: id,
              clientMutationId:  "MDQ6VXNlcjM5ODMzMDQ1"
            };
      this.$apollo.mutate({
        mutation: removingStar,
        variables: {
            input: varStar,
        }
      })
      console.log('removed');
    }
  },
  apollo: {
    user: {
      query: overviewTag
    }
  }
}
</script>

<style scoped>
  .repos {
    border-bottom: 1px rgb(141, 134, 134) solid;
    width: 90%;
    height: 100px;
    justify-content: center;
    align-items: center;
    margin: 5px;
    padding: 5px;
    font-weight: 600;
  }
  .btn {
    background-color: #f0f1f1;
    border-color: #a5a6a7;
  }
  .btn:hover {
    background-color: #ccd2db;
  }
  .repoName {
    font-size: 18px;
    color: rgb(10, 112, 196);
  }
  .topics {
    font-size: 14px;
    color: rgb(70, 71, 71);
  }
</style>
