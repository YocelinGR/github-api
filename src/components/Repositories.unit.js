import Repositories from './Repositories.vue'
import { createLocalVue } from '@vue/test-utils'

describe('@components/Repositories', () => {
  const localVue = createLocalVue()
  it('Repositories is a vue instance', () => {
    const wrapper = shallowMount(Repositories, { localVue })
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('called Apollo addStar mutation', () => {
    const mutate = jest.fn()
    const wrapper = mount(Repositories, {
      localVue,
      mocks: {
        $apollo: {
          mutate,
        },
      },
    })
    wrapper.vm.addStar()
    expect(mutate).toBeCalled()
  })

  it('called Apollo removingStar mutation', () => {
    const mutate = jest.fn()
    const wrapper = mount(Repositories, {
      localVue,
      mocks: {
        $apollo: {
          mutate,
        },
      },
    })
    wrapper.vm.removingStar()
    expect(mutate).toBeCalled()
  })

  it('Displays user starred repositories data', () => {
    const wrapper = shallowMount(Repositories, { localVue })
    wrapper.setData({
      user: {
        id: "MDQ6VXNlcjM5ODMzMDQ1",
        repositories: {
          nodes: [
            {
              id: "MDEwOlJlcG9zaXRvcnkxNzE3NjAwMjE=",
              name: "graphql-subscriptions-server-practice",
              url: "https://github.com/YocelinGR/graphql-subscriptions-server-practice",
              viewerHasStarred: false,
              createdAt: "2019-02-20T22:35:48Z",
              repositoryTopics: {
                nodes: []
              }
            }
          ]
        }
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
