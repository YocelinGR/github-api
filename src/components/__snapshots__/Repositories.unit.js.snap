// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@components/Repositories Displays user starred repositories data 1`] = `
<div>
  <div
    class="repos row"
  >
    <div
      class="col-md-6"
    >
      <a
        class="repoName"
        href="https://github.com/YocelinGR/graphql-subscriptions-server-practice"
      >
        graphql-subscriptions-server-practice
      </a>
       
      <br />
       
      <span
        class="topics"
      />
       
      <span
        class="topics"
      >
        2019-02-20T22:35:48Z
      </span>
    </div>
     
    <div
      class="col-md-3"
    >
      <button
        class="btn btn-light"
      >
        <i
          class="far fa-star"
        />
      </button>
    </div>
  </div>
</div>
`;
